desc/objectives/_geometry.py

```diff
@@ -1393,3 +1393,593 @@ def compute(self, params, constants=None):
+        if `normalize` is `True` and the target is in physical units, this should also
+        be set to True.
+    loss_function : {None, 'mean', 'min', 'max'}, optional
+        Loss function to apply to the objective values once computed. This loss function
+        is called on the raw compute value, before any shifting, scaling, or
+        normalization.
+    deriv_mode : {"auto", "fwd", "rev"}
+        Specify how to compute jacobian matrix, either forward mode or reverse mode AD.
+        "auto" selects forward or reverse mode based on the size of the input and output
+        of the objective. Has no effect on self.grad or self.hess which always use
+        reverse mode and forward over reverse mode respectively.
+    grid : Grid, optional
+        Collocation grid containing the nodes to evaluate at. Defaults to
+        ``LinearGrid(M=eq.M_grid, N=eq.N_grid)`` for ``Equilibrium``
+        or ``LinearGrid(M=2*eq.M, N=2*eq.N)`` for ``FourierRZToroidalSurface``.
+    name : str, optional
+        Name of the objective function.
+
+    """
+
+    _coordinates = "rtz"
+    _units = "(m^-1)"
+    _print_value_fmt = "Umbilic curvature: {:10.3e} "
+
+    def __init__(
+        self,
+        eq,
+        curve,
+        target=None,
+        bounds=None,
+        weight=1,
+        curve_grid=None,
+        equil_grid=None,
+        equil_grid_axis=None,
+        curve_fixed=False,
+        normalize=True,
+        normalize_target=True,
+        loss_function=None,
+        deriv_mode="auto",
+        grid=None,
+        name="Umbilic curvature",
+    ):
+
+        if target is None and bounds is None:
+            target = -10
+
+        self._eq = curve
+        self._curve = curve
+        self._curve_grid = curve_grid
+        self._equil_grid = equil_grid
+        self._equil_grid_axis = equil_grid_axis
+        self._curve_fixed = curve_fixed
+        super().__init__(
+            things=[eq, self._curve] if not curve_fixed else [eq],
+            target=target,
+            bounds=bounds,
+            weight=weight,
+            normalize=normalize,
+            normalize_target=normalize_target,
+            loss_function=loss_function,
+            deriv_mode=deriv_mode,
+            name=name,
+        )
+
+    def build(self, use_jit=True, verbose=1):
+        """Build constant arrays.
+
+        Parameters
+        ----------
+        use_jit : bool, optional
+            Whether to just-in-time compile the objective and derivatives.
+        verbose : int, optional
+            Level of output.
+
+        """
+        eq = self.things[0]
+        curve = self._curve if self._curve_fixed else self.things[1]
+        # if things[1] is different than self._curve, update self._curve
+        if curve != self._curve:
+            self._curve = curve
+        if self._curve_grid is None:
+            phi_arr = jnp.linspace(0, 2 * jnp.pi, 4 * curve.N)
+            phi_arr = jnp.concatenate(
```

Comment:
what is going on here exactly? you ad 2pi and then mod it? does this only work for an umbilic factor of 3?

Reply:
if this needs to go to 2pi * NFP_umbilic_fac, then need to make this more general, not hardcoded for factor of 3